Update description files (#123)
ballerina-platform/module-ballerina-constraint: README.md

  [![GitHub Last Commit](https://img.shields.io/github/last-commit/ballerina-platform/module-ballerina-constraint.svg?label=Last%20Commit)](https://github.com/ballerina-platform/module-ballerina-constraint/commits/master)
  [![GitHub issues](https://img.shields.io/github/issues/ballerina-platform/ballerina-standard-library/module/constraint.svg?label=Open%20Issues)](https://github.com/ballerina-platform/ballerina-standard-library/labels/module%2Fconstraint)

The Ballerina `constraint` library facilitates APIs to do validations on the Ballerina types further with the use of annotations.

### Constraint annotations

> This library provides the following annotations to validate the values that have been assigned to Ballerina types.

Reviewer: ```suggestion
This library provides the following annotations to validate the values that have been set for Ballerina types.
```

| `int`                             | `@constraint:Int`    | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |
| `float`                           | `@constraint:Float`  | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |
| `int`&#124;`float`&#124;`decimal` | `@constraint:Number` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |
| `string`                          | `@constraint:String` | `length`, `minLength`, `maxLength`, `pattern`                                                                          |
| `any[]`                           | `@constraint:Array`  | `length`, `minLength`, `maxLength`                                                                                     |
| `constraint:Date`                 | `@constraint:Date`   | `option` - `PAST` or `FUTURE` or `PAST_OR_PRESENT` or `FUTURE_OR_PRESENT`                                              |

> The following example demonstrates how to apply constraint annotations to Ballerina types.

Reviewer: ```suggestion
The following example demonstrates how to apply constraint annotations to types.
```

```ballerina
@constraint:String { pattern: re `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$` }
public type Email string;

@constraint:Date { option: constraint:PAST }
public type DOB time:Date;

public type Person record {|
    @constraint:String { minLength: 5, maxLength: 10 }
    string name;
    @constraint:Int { minValue: 18, maxValue: 60 }
    int age;
    Email email;
    @constraint:Array { minLength: 1, maxLength: 5 }
    string[] phoneNumbers;
    DOB dob;
|};
```

### Constraint validation

> The `validate` function in this library can be used to validate the values that have been assigned to Ballerina types with respect to the constraints defined using the annotations.

Reviewer: Would change similarly. Since we aren't assigning values to type, but creating values of those types or providing values belonging to those types.

Author: Fixed via [d30a43e](https://github.com/ballerina-platform/module-ballerina-constraint/pull/123/commits/d30a43eca8463f89ed28fd35c2f6ee3d84b0b4c2)

  [![GitHub Last Commit](https://img.shields.io/github/last-commit/ballerina-platform/module-ballerina-constraint.svg?label=Last%20Commit)](https://github.com/ballerina-platform/module-ballerina-constraint/commits/master)
  [![GitHub issues](https://img.shields.io/github/issues/ballerina-platform/ballerina-standard-library/module/constraint.svg?label=Open%20Issues)](https://github.com/ballerina-platform/ballerina-standard-library/labels/module%2Fconstraint)

> The Ballerina `constraint` library provides annotations to add constraints to Ballerina types and API to validate the values with respect to the constraints defined in the respective types.

Reviewer: ```suggestion
The Ballerina `constraint` library provides annotations to add constraints to Ballerina types and an API to validate the values with respect to the constraints defined in the respective types.
```

or

```suggestion
The Ballerina `constraint` library provides annotations to add constraints to Ballerina types and a function to validate the values with respect to the constraints defined in the respective types.
```

Author: Using API rather than function

Fixed via [9c58851](https://github.com/ballerina-platform/module-ballerina-constraint/pull/123/commits/9c58851bfdca933e13b09490135ee6b9b64025de)

### Constraint annotations

This library provides the following annotations on Ballerina types to validate the values created with the respective types.

| Ballerina Type                    | Annotation           | Supported Constraints                                                                                                  |
|-----------------------------------|----------------------|------------------------------------------------------------------------------------------------------------------------|
| `int`                             | `@constraint:Int`    | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |
| `float`                           | `@constraint:Float`  | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |

> | `int`&#124;`float`&#124;`decimal` | `@constraint:Number` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` |

Reviewer 2: Why do we need to specifically mention the `int|float|decimal` union type in here? And if we do so, we would have to mention the other supported union type scenarios as well right?

Author: Here the type denote the effective type which supports these annotations. Since number constraints are allowed to all number types, I have added it as a union of Ballerina (unique) number types. The other constraints are applied to only one effective type

Reviewer 2: I see. maybe I got distracted because I was focusing on the types rather than the annotations themselves. Ideally, the table should be structured around the annotations. One suggestion I have is to revise the table by making the annotations as first column and, then include the second column titled "Associated Ballerina Type(s)" to clarify that this section is about annotations, not the types themselves. This would help emphasize the main point of the section. wdyt?

Author: +1, will change

Author: Fixed via [84179f4](https://github.com/ballerina-platform/module-ballerina-constraint/pull/123/commits/84179f42d6e2dd50ce86dda55d71e439f1a0c03c)

|----------------------|------------------------------------------------------------------------------------------------------------------------|------------------------------|
| `@constraint:Int`    | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` | `int`                        |
| `@constraint:Float`  | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` | `float`                      |
| `@constraint:Number` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive`, `maxDigits`, `maxIntegerDigits`, `maxFractionDigits` | `int`, `float`, `decimal`    |
| `@constraint:String` | `length`, `minLength`, `maxLength`, `pattern`                                                                          | `string`                     |
| `@constraint:Array`  | `length`, `minLength`, `maxLength`                                                                                     | `any[]`                      |

> | `@constraint:Date`   | `option` - `PAST` or `FUTURE` or `PAST_OR_PRESENT` or `FUTURE_OR_PRESENT`                                              | `constraint:Date`            |

Reviewer 2: Shouldn't this be
```suggestion
| `@constraint:Date`   | `option` - `PAST` or `FUTURE` or `PAST_OR_PRESENT` or `FUTURE_OR_PRESENT`                                              | `time:Date`            |
```
?

Author: The `time:Date` some additional (optional) fields which are not required for the validation. So we have defined a simple record including the necessary fields in the `constraint` package, which makes it independent of the time package (this was a [design decision](https://github.com/ballerina-platform/ballerina-library/issues/3960#issuecomment-1550877300)). Also `time:Date` is always `constraint:Date`.

Reviewer 2: Got it. makes sense.
